sndio.c

```diff
@@ -0,0 +1,268 @@
+{
+	sndio_volume = l > r ? l : r;
+
+	if (hdl == NULL)
+		return -OP_ERROR_NOT_INITIALIZED;
+
+	if (!sio_setvol(hdl, sndio_volume * SIO_MAXVOL / 100))
+		return -OP_ERROR_INTERNAL;
+
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_mixer_get_volume(int *l, int *r)
+{
+	*l = *r = sndio_volume;
+
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_set_sf(sample_format_t sf)
+{
+	struct sio_par apar;
+
+	sndio_sf = sf;
+
+	sio_initpar(&par);
+
+	par.pchan = sf_get_channels(sndio_sf);
+	par.rate = sf_get_rate(sndio_sf);
+	sndio_paused = 0;
+
+	if (sf_get_signed(sndio_sf))
+		par.sig = 1;
+	else
+		par.sig = 0;
+
+	if (sf_get_bigendian(sndio_sf))
+		par.le = 0;
+	else
+		par.le = 1;
+
+	switch (sf_get_bits(sndio_sf)) {
+	case 16:
+		par.bits = 16;
+		break;
+	case 8:
+		par.bits = 8;
+		break;
+	default:
+		return -OP_ERROR_SAMPLE_FORMAT;
+	}
+
+	par.appbufsz = par.rate * 300 / 1000;
+	apar = par;
+
+	if (!sio_setpar(hdl, &par))
+		return -OP_ERROR_INTERNAL;
+
+	if (!sio_getpar(hdl, &par))
+		return -OP_ERROR_INTERNAL;
+
+	if (apar.rate != par.rate || apar.pchan != par.pchan ||
+	    apar.bits != par.bits || (par.bits > 8 && apar.le != par.le) ||
+	    apar.sig != par.sig)
+		return -OP_ERROR_INTERNAL;
+
+	sndio_mixer_set_volume(sndio_volume, sndio_volume);
+
+	if (!sio_start(hdl))
+		return -OP_ERROR_INTERNAL;
+
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_init(void)
+{
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_exit(void)
+{
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_close(void)
+{
+	if (hdl != NULL) {
+		sio_close(hdl);
+		hdl = NULL;
+	}
+
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_open(sample_format_t sf, const channel_position_t *channel_map)
+{
+	int ret = 0;
+
+	hdl = sio_open(NULL, SIO_PLAY, 0);
+	if (hdl == NULL)
+		return -OP_ERROR_INTERNAL;
+
+	if ((ret = sndio_set_sf(sf)) < 0) {
+		sndio_close();
+		return ret;
+	}
+
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_write(const char *buf, int cnt)
+{
+	size_t rc;
+
+	rc = sio_write(hdl, buf, cnt);
+	if (rc == 0)
+		return -OP_ERROR_INTERNAL;
+
+	return rc;
+}
+
+static int op_sndio_set_option(int key, const char *val)
+{
+	return -OP_ERROR_NOT_OPTION;
+}
+
+static int op_sndio_get_option(int key, char **val)
+{
+	return -OP_ERROR_NOT_OPTION;
+}
+
+static int sndio_pause(void)
+{
+	if (!sndio_paused) {
+		if(!sio_stop(hdl))
+			return -OP_ERROR_INTERNAL;
+		sndio_paused = 1;
+	}
+
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_unpause(void)
+{
+	if (sndio_paused) {
+		if (!sio_start(hdl))
+			return -OP_ERROR_INTERNAL;
+		sndio_paused = 0;
+	}
+
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_buffer_space(void)
+{
+	/*
+	 * Do as if there's always some space and let sio_write() block.
+	 */
+	return par.bufsz * par.bps * par.pchan;
+}
+
+static int sndio_mixer_init(void)
+{
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_mixer_exit(void)
+{
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_mixer_open(int *volume_max)
+{
+	*volume_max = 100;
```

Comment:
Sorry, Github hid the comments because diffs were «outdated» — so I missed this note.

`…_mixer_open(int *volume_max)` has `volume_max` precisely so that mixer plugins do not have to do any scaling internally. I don't think this is a breaking change for users, though. I'll change this myself before merging for code to be a bit cleaner.